File: Springboot_API/RetaCantabria/src/main/java/org/example/Entidades/PuntoRuta.java
package org.example.Entidades;

import com.fasterxml.jackson.annotation.JsonBackReference;
import jakarta.persistence.*;

import java.time.LocalDateTime;

@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "tipo_punto")
public abstract class PuntoRuta {
    @Id
    @GeneratedValue(strategy= GenerationType.IDENTITY)
    private Integer idPuntoRuta;

    private double longitud;
    private double latitud;
    private int elevacion;
    private LocalDateTime timestamp;

    @ManyToOne(optional = false)
    @JoinColumn(name = "idRuta")
    @JsonBackReference
    private Ruta ruta;

    public PuntoRuta() {
    }

    public PuntoRuta(int idPuntoRuta, double longitud, double latitud, int elevacion, LocalDateTime timestamp, Ruta ruta) {
        this.idPuntoRuta = idPuntoRuta;
        this.longitud = longitud;
        this.latitud = latitud;
        this.elevacion = elevacion;
        this.timestamp = timestamp;
        this.ruta = ruta;
    }

    public Integer getIdPuntoRuta() {
        return idPuntoRuta;
    }

    public void setIdPuntoRuta(int idPuntoRuta) {
        this.idPuntoRuta = idPuntoRuta;
    }

    public double getLongitud() {
        return longitud;
    }

    public void setLongitud(double longitud) {
        this.longitud = longitud;
    }

    public double getLatitud() {
        return latitud;
    }

    public void setLatitud(double latitud) {
        this.latitud = latitud;
    }

    public int getElevacion() {
        return elevacion;
    }

    public void setElevacion(int elevacion) {
        this.elevacion = elevacion;
    }

    public LocalDateTime getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(LocalDateTime timestamp) {
        this.timestamp = timestamp;
    }

    public Ruta getRuta() {
        return ruta;
    }

    public void setRuta(Ruta ruta) {
        this.ruta = ruta;
    }

    public abstract String getTipo();
}

File: AplicacionMovil/app/src/main/java/com/example/retodam2rutas/entities/PuntoRuta.kt
package com.example.retodam2rutas.entities

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey
import java.time.LocalDateTime

@Entity(
    tableName = "puntosRuta",
    foreignKeys = [
        ForeignKey(
            entity = Ruta::class,
            parentColumns = ["idRuta"],
            childColumns = ["rutaId"],
            onDelete = ForeignKey.CASCADE
        )
    ],
    indices = [Index("rutaId")]
)
data class PuntoRuta(

    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    val id: Long = 0L,

    @ColumnInfo(name = "latitud")
    val latitud: Double,

    @ColumnInfo(name = "longitud")
    val longitud: Double,

    @ColumnInfo(name = "elevacion")
    val elevacion: Int,

    @ColumnInfo(name = "timestamp")
    val timeStamp: LocalDateTime,

    @ColumnInfo(name = "rutaId")
    val rutaId: Long? = null
)

